types/index.d.ts

```diff
@@ -0,0 +1,329 @@
+
+	// Standard, but not in the TS lib.
+	/** Returns or sets a `DOMMatrix` for the current transformation matrix. */
+	currentTransform: NodeCanvasDOMMatrix
+
+	// Standard, but need node-canvas class versions:
+	getTransform(): NodeCanvasDOMMatrix
+	setTransform(a: number, b: number, c: number, d: number, e: number, f: number): void
+	setTransform(transform?: NodeCanvasDOMMatrix): void
+	createImageData(sw: number, sh: number): NodeCanvasImageData;
+	createImageData(imagedata: NodeCanvasImageData): NodeCanvasImageData;
+	getImageData(sx: number, sy: number, sw: number, sh: number): NodeCanvasImageData;
+	putImageData(imagedata: NodeCanvasImageData, dx: number, dy: number): void;
+	putImageData(imagedata: NodeCanvasImageData, dx: number, dy: number, dirtyX: number, dirtyY: number, dirtyWidth: number, dirtyHeight: number): void;
+	drawImage(image: Canvas|Image, dx: number, dy: number): void;
+	drawImage(image: Canvas|Image, dx: number, dy: number, dw: number, dh: number): void;
+	drawImage(image: Canvas|Image, sx: number, sy: number, sw: number, sh: number, dx: number, dy: number, dw: number, dh: number): void;
+	/**
+	 * **Do not use this overload. Use one of the other three overloads.** This
+	 * is a catch-all definition required for compatibility with the base
+	 * `CanvasRenderingContext2D` interface.
+	 */
+	drawImage(...args: any[]): void;
+}
+export { NodeCanvasRenderingContext2D as CanvasRenderingContext2D }
+
+declare class NodeCanvasCanvasGradient extends CanvasGradient {}
+export { NodeCanvasCanvasGradient as CanvasGradient }
+
+declare class NodeCanvasCanvasPattern extends CanvasPattern {}
+export { NodeCanvasCanvasPattern as CanvasPattern }
+
+// This does not extend HTMLImageElement because there are dozens of inherited
+// methods and properties that we do not provide.
+export class Image {
+	/** Track image data */
+	static readonly MODE_IMAGE: number
+	/** Track MIME data */
+	static readonly MODE_MIME: number
+
+	/**
+	 * The URL, `data:` URI or local file path of the image to be loaded, or a
+	 * Buffer instance containing an encoded image.
+	 */
+	src: string | Buffer
+	/** Retrieves whether the object is fully loaded. */
+	readonly complete: boolean
+	/** Sets or retrieves the height of the image. */
+	height: number
+	/** Sets or retrieves the width of the image. */
+	width: number
+
+	/** The original height of the image resource before sizing. */
+	readonly naturalHeight: number;
+	/** The original width of the image resource before sizing. */
+	readonly naturalWidth: number;
+	/**
+	 * Applies to JPEG images drawn to PDF canvases only. Setting
+	 * `img.dataMode = Image.MODE_MIME` or `Image.MODE_MIME|Image.MODE_IMAGE`
+	 * enables image MIME data tracking. When MIME data is tracked, PDF canvases
+	 * can embed JPEGs directly into the output, rather than re-encoding into
+	 * PNG. This can drastically reduce filesize and speed up rendering.
+	 */
+	dataMode: number
+
+	onload(): void
+	onerror(err: Error): void
```

Comment:
Since you can set and unset the handlers, they should be marked as optional:

```diff
-	onload(): void
-	onerror(err: Error): void
+	onload?(): void
+	onerror?(err: Error): void
```

Reply:
Will make the declaration fix, thanks.

That shouldn't be necessary to avoid leaks since https://github.com/Automattic/node-canvas/pull/1003. If you can repro a leak with the listeners, can you please open an issue for it?

Reply:
Thanks for the link, I wasn't aware that `node-canvas` can avoid these kinds of leaks. I still prefer to manually unset the handler: not all libraries handle this properly so I prefer to always do it rather than remembering which libraries are smarter.

By the way: I always use the documented handlers but does `Image` supports an event-emitter based API? (`addEventListener`, `removeEventListener`, etc.)? If so, it should also be typed.

Reply:
Good question -- no, that's #1218. The only reason I haven't worked on that is because I'm not sure if we should (a) use the Node.js EventEmitter API (`on()`, `addListener()`, `off()`, `removeListener()`), (b) use the HTML API (`addEventListener()`, `removeEventListener()`), or (c) use [a] along with aliases for [b]. (If you have any thoughts on it, feel free to comment in the issue!)